invenio_app_ils/ui/src/invenio_app_ils/common/api/loans/loan.js

```diff
@@ -0,0 +1,116 @@
+import { http } from '../base';
+import { toISO } from '../date';
+import { DateTime } from 'luxon';
+
+const loanURL = '/circulation/loans/';
+const get = loanPid => http.get(`${loanURL}${loanPid}`);
+const getLoanReplaceItemUrl = loanPid => `${loanURL}/${loanPid}/replace-item`;
+
+const postAction = (
+  url,
+  pid,
+  loan,
+  transactionUserPid,
+  transactionLocationPid
+) => {
+  const now = DateTime.local();
+  const payload = {
+    transaction_user_pid: transactionUserPid,
+    patron_pid: loan.patron_pid,
+    transaction_location_pid: transactionLocationPid,
+    transaction_date: toISO(now),
+  };
+
+  if ('item_pid' in loan) {
+    payload['item_pid'] = loan.item_pid;
+  } else if ('document_pid' in loan) {
+    payload['document_pid'] = loan.document_pid;
+  } else {
+    throw new Error(
+      `No 'item_pid' or 'document_pid' attached to loan '${pid}'`
+    );
+  }
+  return http.post(url, payload);
+};
+
+const assignItemToLoan = (itemPid, loanPid) => {
+  const url = getLoanReplaceItemUrl(loanPid);
+  const payload = { item_pid: itemPid };
+  return http.post(url, payload);
+};
+
+class Loans {
+  constructor(build) {
+    this.documentQuery = build.documentQuery ? build.documentQuery : '';
+    this.itemQuery = build.itemQuery ? build.itemQuery : '';
+    this.patronQuery = build.patronQuery ? build.patronQuery : '';
+    this.stateQuery = build.stateQuery ? build.stateQuery : '';
+  }
+
+  qs() {
+    return `${this.documentQuery}${this.itemQuery}${this.patronQuery}${
+      this.stateQuery
+    }`.replace(' AND ', '');
+  }
+
+  static get Builder() {
+    class Builder {
+      withDocPid(documentPid) {
+        this.documentQuery = documentPid
+          ? ` AND document_pid:${Builder.checkSum(documentPid)}`
+          : '';
+        return this;
+      }
+
+      withItemPid(itemPid) {
+        this.itemQuery = itemPid
+          ? ` AND item_pid:${Builder.checkSum(itemPid)}`
+          : '';
+        return this;
+      }
+
+      withPatronPid(patronPid) {
+        this.patronQuery = patronPid
+          ? ` AND patron_pid:${Builder.checkSum(patronPid)}`
+          : '';
+        return this;
+      }
+
+      withState(state) {
+        this.stateQuery = state ? ` AND state:${Builder.checkSum(state)}` : '';
+        return this;
+      }
+
+      static checkSum(param) {
+        if (Array.isArray(param)) {
+          const paramQuery = param.join(' OR ');
+          return `(${paramQuery})`;
+        } else {
+          return param;
+        }
+      }
+
+      build() {
+        return new Loans(this).qs();
+      }
+    }
+    return Builder;
+  }
+}
+
+const query = () => {
```

Comment:
I would personally call it with `builder` somewhere, to state that you actually have to build your query. `queryBuilder`?